Parse succeeds when transaction is in error state

**Describe the problem**

Standard PostgreSQL returns an `ErrorResponse` in response to `Parse` when in a transaction in an error state. CockroachDB succeeds (or at least it appears to). 

**To Reproduce**

Run this test in the pgx test suite: https://github.com/jackc/pgx/blob/396195466c6a0cd445d3690a3e33551d7808cf85/conn_test.go#L1067


**Expected behavior**

`ErrorResponse` should be returned.

**Additional data / screenshots**

Trace log when running against PG 14:

```
jack@glados ~/dev/pgx ±v5-dev⚡ » go test -run TestStmtCacheInvalidationTx                                      1 ↵
Alias tip: got -run TestStmtCacheInvalidationTx
F	Query	196	 "
        DROP TABLE IF EXISTS drop_cols;
        CREATE TABLE drop_cols (
            id SERIAL PRIMARY KEY NOT NULL,
            f1 int NOT NULL,
            f2 int NOT NULL
        );
    "
B	CommandComplete	16	 "DROP TABLE"
B	CommandComplete	18	 "CREATE TABLE"
B	ReadyForQuery	6	 I
F	Query	50	 "INSERT INTO drop_cols (f1, f2) VALUES (1, 2)"
B	CommandComplete	16	 "INSERT 0 1"
B	ReadyForQuery	6	 I
F	Query	11	 "begin"
B	CommandComplete	11	 "BEGIN"
B	ReadyForQuery	6	 T
F	Parse	57	 "stmtcache_1" "SELECT * FROM drop_cols WHERE id = $1" 0
F	Describe	18	 S "stmtcache_1"
F	Sync	5
B	ParseComplete	5
B	ParameterDescription	11
B	RowDescription	70	 3 "id" 21620847 1 23 4 -1 0 "f1" 21620847 2 23 4 -1 0 "f2" 21620847 3 23 4 -1 0
B	ReadyForQuery	6	 T
F	Bind	40	 "" "stmtcache_1" 1 1 1 '\x0\x0\x0\x1' 3 1 1 1
F	Describe	7	 P ""
F	Execute	10	 "" 0
F	Sync	5
B	BindComplete	5
B	RowDescription	70	 3 "id" 21620847 1 23 4 -1 1 "f1" 21620847 2 23 4 -1 1 "f2" 21620847 3 23 4 -1 1
B	DataRow	31	 3 4 '\x0\x0\x0\x1' 4 '\x0\x0\x0\x1' 4 '\x0\x0\x0\x2'
B	CommandComplete	14	 "SELECT 1"
B	ReadyForQuery	6	 T
F	Query	42	 "ALTER TABLE drop_cols DROP COLUMN f1"
B	CommandComplete	17	 "ALTER TABLE"
B	ReadyForQuery	6	 T
F	Bind	40	 "" "stmtcache_1" 1 1 1 '\x0\x0\x0\x1' 3 1 1 1
F	Describe	7	 P ""
F	Execute	10	 "" 0
F	Sync	5
B	ErrorResponse	109
B	ReadyForQuery	6	 E
F	Parse	57	 "stmtcache_2" "SELECT * FROM drop_cols WHERE id = $1" 0
F	Describe	18	 S "stmtcache_2"
F	Sync	5
B	ErrorResponse	146
B	ReadyForQuery	6	 E
F	Query	14	 "rollback"
B	CommandComplete	14	 "ROLLBACK"
B	ReadyForQuery	6	 I
F	Close	18
F	Sync	5
B	CloseComplete	5
B	ReadyForQuery	6	 I
F	Parse	57	 "stmtcache_3" "SELECT * FROM drop_cols WHERE id = $1" 0
F	Describe	18	 S "stmtcache_3"
F	Sync	5
B	ParseComplete	5
B	ParameterDescription	11
B	RowDescription	70	 3 "id" 21620847 1 23 4 -1 0 "f1" 21620847 2 23 4 -1 0 "f2" 21620847 3 23 4 -1 0
B	ReadyForQuery	6	 I
F	Bind	40	 "" "stmtcache_3" 1 1 1 '\x0\x0\x0\x1' 3 1 1 1
F	Describe	7	 P ""
F	Execute	10	 "" 0
F	Sync	5
B	BindComplete	5
B	RowDescription	70	 3 "id" 21620847 1 23 4 -1 1 "f1" 21620847 2 23 4 -1 1 "f2" 21620847 3 23 4 -1 1
B	DataRow	31	 3 4 '\x0\x0\x0\x1' 4 '\x0\x0\x0\x1' 4 '\x0\x0\x0\x2'
B	CommandComplete	14	 "SELECT 1"
B	ReadyForQuery	6	 I
F	Parse	48	 "stmtcache_4" "select generate_series(1,$1)" 0
F	Describe	18	 S "stmtcache_4"
F	Sync	5
B	ParseComplete	5
B	ParameterDescription	11
B	RowDescription	41	 1 "generate_series" 0 0 23 4 -1 0
B	ReadyForQuery	6	 I
F	Bind	36	 "" "stmtcache_4" 1 1 1 '\x0\x0\x0\xa' 1 1
F	Describe	7	 P ""
F	Execute	10	 "" 0
F	Sync	5
B	BindComplete	5
B	RowDescription	41	 1 "generate_series" 0 0 23 4 -1 1
B	DataRow	15	 1 4 '\x0\x0\x0\x1'
B	DataRow	15	 1 4 '\x0\x0\x0\x2'
B	DataRow	15	 1 4 '\x0\x0\x0\x3'
B	DataRow	15	 1 4 '\x0\x0\x0\x4'
B	DataRow	15	 1 4 '\x0\x0\x0\x5'
B	DataRow	15	 1 4 '\x0\x0\x0\x6'
B	DataRow	15	 1 4 '\x0\x0\x0\x7'
B	DataRow	15	 1 4 '\x0\x0\x0\x8'
B	DataRow	15	 1 4 '\x0\x0\x0\x9'
B	DataRow	15	 1 4 '\x0\x0\x0\xa'
B	CommandComplete	15	 "SELECT 10"
B	ReadyForQuery	6	 I
F	Terminate	5
PASS
ok  	github.com/jackc/pgx/v5	0.186s
```

Trace log when running against CRDB:

```
jack@glados ~/dev/pgx ±v5-dev⚡ » PGX_TEST_DATABASE='postgresql://root@127.0.0.1:26257/pgx_test?sslmode=disable&experimental_enable_temp_tables=on' go test -run TestStmtCacheInvalidationTx
F	Query	196	 "
        DROP TABLE IF EXISTS drop_cols;
        CREATE TABLE drop_cols (
            id SERIAL PRIMARY KEY NOT NULL,
            f1 int NOT NULL,
            f2 int NOT NULL
        );
    "
B	CommandComplete	16	 "DROP TABLE"
B	NoticeResponse	201
B	CommandComplete	18	 "CREATE TABLE"
B	ReadyForQuery	6	 I
F	Query	50	 "INSERT INTO drop_cols (f1, f2) VALUES (1, 2)"
B	CommandComplete	16	 "INSERT 0 1"
B	ReadyForQuery	6	 I
F	Query	11	 "begin"
B	CommandComplete	11	 "BEGIN"
B	ReadyForQuery	6	 T
F	Parse	57	 "stmtcache_1" "SELECT * FROM drop_cols WHERE id = $1" 0
F	Describe	18	 S "stmtcache_1"
F	Sync	5
B	ParseComplete	5
B	ParameterDescription	11
B	RowDescription	70	 3 "id" 115 1 20 8 -1 0 "f1" 115 2 20 8 -1 0 "f2" 115 3 20 8 -1 0
B	ReadyForQuery	6	 T
F	Bind	44	 "" "stmtcache_1" 1 1 1 '\x0\x0\x0\x0\x0\x0\x0\x1' 3 1 1 1
F	Describe	7	 P ""
F	Execute	10	 "" 0
F	Sync	5
B	BindComplete	5
B	RowDescription	70	 3 "id" 115 1 20 8 -1 1 "f1" 115 2 20 8 -1 1 "f2" 115 3 20 8 -1 1
B	CommandComplete	14	 "SELECT 0"
B	ReadyForQuery	6	 T
F	Query	42	 "ALTER TABLE drop_cols DROP COLUMN f1"
B	CommandComplete	17	 "ALTER TABLE"
B	ReadyForQuery	6	 T
F	Bind	44	 "" "stmtcache_1" 1 1 1 '\x0\x0\x0\x0\x0\x0\x0\x1' 3 1 1 1
F	Describe	7	 P ""
F	Execute	10	 "" 0
F	Sync	5
B	BindComplete	5
B	RowDescription	70	 3 "id" 115 1 20 8 -1 1 "f1" 115 2 20 8 -1 1 "f2" 115 3 20 8 -1 1
B	ErrorResponse	95
B	ReadyForQuery	6	 E
F	Parse	57	 "stmtcache_2" "SELECT * FROM drop_cols WHERE id = $1" 0
F	Describe	18	 S "stmtcache_2"
F	Sync	5
B	ParseComplete	5
B	ParameterDescription	11
B	RowDescription	49	 2 "id" 115 1 20 8 -1 0 "f2" 115 3 20 8 -1 0
B	ReadyForQuery	6	 E
F	Bind	42	 "" "stmtcache_2" 1 1 1 '\x0\x0\x0\x0\x0\x0\x0\x1' 2 1 1
F	Describe	7	 P ""
F	Execute	10	 "" 0
F	Sync	5
B	BindComplete	5
B	RowDescription	49	 2 "id" 115 1 20 8 -1 1 "f2" 115 3 20 8 -1 1
B	ErrorResponse	144
B	ReadyForQuery	6	 E
F	Query	14	 "rollback"
B	CommandComplete	14	 "ROLLBACK"
B	ReadyForQuery	6	 I
F	Close	18
F	Sync	5
B	CloseComplete	5
B	ReadyForQuery	6	 I
F	Bind	42	 "" "stmtcache_2" 1 1 1 '\x0\x0\x0\x0\x0\x0\x0\x1' 2 1 1
F	Describe	7	 P ""
F	Execute	10	 "" 0
F	Sync	5
B	BindComplete	5
B	RowDescription	49	 2 "id" 115 1 20 8 -1 1 "f2" 115 3 20 8 -1 1
B	ErrorResponse	95
B	ReadyForQuery	6	 I
F	Terminate	5
--- FAIL: TestStmtCacheInvalidationTx (0.94s)
    conn_test.go:1043:
        	Error Trace:	conn_test.go:1043
        	Error:      	Received unexpected error:
        	            	ERROR: cached plan must not change result type (SQLSTATE 0A000)
        	Test:       	TestStmtCacheInvalidationTx
FAIL
exit status 1
FAIL	github.com/jackc/pgx/v5	1.115s
```


**Environment:**

```
jack@glados ~/dev/pgx ±v5-dev⚡ » cockroach --version
cockroach version details:
Build Tag:        v22.1.2
Build Time:       2022/06/22 16:09:32
Distribution:     CCL
Platform:         darwin amd64 (x86_64-apple-darwin19)
Go Version:       go1.17.11
C Compiler:       Clang 10.0.0
Build Commit ID:  fc456a26830067b6dfbb8cd87e093a28d9b833d1
Build Type:       release
```

**Additional context**
What was the impact?

This behavior can confuse pgx's automatic statement caching and invalidation system.


Jira issue: CRDB-17488
Epic CRDB-14049

--- a/pkg/sql/conn_executor_prepare.go
+++ b/pkg/sql/conn_executor_prepare.go
@@ -19,6 +19,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgwirebase"
 	"github.com/cockroachdb/cockroach/pkg/sql/querycache"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlerrors"
 	"github.com/cockroachdb/cockroach/pkg/sql/types"
 	"github.com/cockroachdb/cockroach/pkg/util/fsm"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
@@ -40,6 +41,10 @@ func (ex *connExecutor) execPrepare(
 	// the Sync message is handled.
 	if _, isNoTxn := ex.machine.CurState().(stateNoTxn); isNoTxn {
 		return ex.beginImplicitTxn(ctx, parseCmd.AST)
+	} else if _, isAbortedTxn := ex.machine.CurState().(stateAborted); isAbortedTxn {
+		if !ex.isAllowedInAbortedTxn(parseCmd.AST) {
+			return retErr(sqlerrors.NewTransactionAbortedError("" /* customMsg */))
+		}
 	}
 
 	ctx, sp := tracing.EnsureChildSpan(ctx, ex.server.cfg.AmbientCtx.Tracer, "prepare stmt")
@@ -305,6 +310,10 @@ func (ex *connExecutor) execBind(
 	// handled separately in conn_executor_exec.
 	if _, isNoTxn := ex.machine.CurState().(stateNoTxn); isNoTxn {
 		return ex.beginImplicitTxn(ctx, ps.AST)
+	} else if _, isAbortedTxn := ex.machine.CurState().(stateAborted); isAbortedTxn {
+		if !ex.isAllowedInAbortedTxn(ps.AST) {
+			return retErr(sqlerrors.NewTransactionAbortedError("" /* customMsg */))
+		}
 	}
 
 	portalName := bindCmd.PortalName
@@ -545,6 +554,7 @@ func (ex *connExecutor) execDescribe(
 	retErr := func(err error) (fsm.Event, fsm.EventPayload) {
 		return eventNonRetriableErr{IsCommit: fsm.False}, eventNonRetriableErrPayload{err: err}
 	}
+	_, isAbortedTxn := ex.machine.CurState().(stateAborted)
 
 	switch descCmd.Type {
 	case pgwirebase.PrepareStatement:
@@ -554,8 +564,6 @@ func (ex *connExecutor) execDescribe(
 				pgcode.InvalidSQLStatementName,
 				"unknown prepared statement %q", descCmd.Name))
 		}
-
-		res.SetInferredTypes(ps.InferredTypes)
 
 		ast := ps.AST
 		if execute, ok := ast.(*tree.Execute); ok {
@@ -570,6 +578,10 @@ func (ex *connExecutor) execDescribe(
 			}
 			ast = innerPs.AST
 		}
+		if isAbortedTxn && !ex.isAllowedInAbortedTxn(ast) {
+			return retErr(sqlerrors.NewTransactionAbortedError("" /* customMsg */))
+		}
+		res.SetInferredTypes(ps.InferredTypes)
 		if stmtHasNoData(ast) {
 			res.SetNoDataRowDescription()
 		} else {
@@ -584,13 +596,20 @@ func (ex *connExecutor) execDescribe(
 				return retErr(pgerror.Newf(
 					pgcode.InvalidCursorName, "unknown portal %q", descCmd.Name))
 			}
+			if isAbortedTxn {
+				return retErr(sqlerrors.NewTransactionAbortedError("" /* customMsg */))
+			}
 			// Sending a nil formatCodes is equivalent to sending all text format
 			// codes.
 			res.SetPortalOutput(ctx, cursor.InternalRows.Types(), nil /* formatCodes */)
 			return nil, nil
 		}
 
-		if stmtHasNoData(portal.Stmt.AST) {
+		ast := portal.Stmt.AST
+		if isAbortedTxn && !ex.isAllowedInAbortedTxn(ast) {
+			return retErr(sqlerrors.NewTransactionAbortedError("" /* customMsg */))
+		}
+		if stmtHasNoData(ast) {
 			res.SetNoDataRowDescription()
 		} else {
 			res.SetPortalOutput(ctx, portal.Stmt.Columns, portal.OutFormats)
@@ -600,4 +619,20 @@ func (ex *connExecutor) execDescribe(
 			"unknown describe type: %s", errors.Safe(descCmd.Type)))
 	}
 	return nil, nil
+}
+
+// isAllowedInAbortedTxn returns true if the statement is allowed to be
+// prepared and executed inside of an aborted transaction.
+func (ex *connExecutor) isAllowedInAbortedTxn(ast tree.Statement) bool {
+	switch s := ast.(type) {
+	case *tree.CommitTransaction, *tree.RollbackTransaction, *tree.RollbackToSavepoint:
+		return true
+	case *tree.Savepoint:
+		if ex.isCommitOnReleaseSavepoint(s.Name) {
+			return true
+		}
+		return false
+	default:
+		return false
+	}
 }

--- a/pkg/sql/pgwire/testdata/pgtest/aborted_txn
+++ b/pkg/sql/pgwire/testdata/pgtest/aborted_txn
@@ -1,0 +1,238 @@
+# Prepare a statement and make sure it works.
+
+send
+Query {"String": "DROP TABLE IF EXISTS drop_cols;"}
+Query {"String": "CREATE TABLE drop_cols (id int PRIMARY KEY NOT NULL, f1 int NOT NULL, f2 int NOT NULL);"}
+Query {"String": "INSERT INTO drop_cols (id, f1, f2) VALUES (1, 1, 2)"}
+Query {"String": "BEGIN"}
+Parse {"Name": "s1", "Query": "SELECT * FROM drop_cols WHERE id = $1"}
+Bind {"PreparedStatement": "s1", "Parameters": [{"text": "1"}]}
+Execute
+Sync
+----
+
+until ignore=NoticeResponse
+ReadyForQuery
+ReadyForQuery
+ReadyForQuery
+ReadyForQuery
+ReadyForQuery
+----
+{"Type":"CommandComplete","CommandTag":"DROP TABLE"}
+{"Type":"ReadyForQuery","TxStatus":"I"}
+{"Type":"CommandComplete","CommandTag":"CREATE TABLE"}
+{"Type":"ReadyForQuery","TxStatus":"I"}
+{"Type":"CommandComplete","CommandTag":"INSERT 0 1"}
+{"Type":"ReadyForQuery","TxStatus":"I"}
+{"Type":"CommandComplete","CommandTag":"BEGIN"}
+{"Type":"ReadyForQuery","TxStatus":"T"}
+{"Type":"ParseComplete"}
+{"Type":"BindComplete"}
+{"Type":"DataRow","Values":[{"text":"1"},{"text":"1"},{"text":"2"}]}
+{"Type":"CommandComplete","CommandTag":"SELECT 1"}
+{"Type":"ReadyForQuery","TxStatus":"T"}
+
+# Make a schema change that breaks the prepared statement.
+
+send
+Query {"String": "ALTER TABLE drop_cols DROP COLUMN f1"}
+Bind {"PreparedStatement": "s1", "Parameters": [{"text": "1"}]}
+Execute
+Sync
+----
+
+until ignore=BindComplete
+ReadyForQuery
+ErrorResponse
+ReadyForQuery
+----
+{"Type":"CommandComplete","CommandTag":"ALTER TABLE"}
+{"Type":"ReadyForQuery","TxStatus":"T"}
+{"Type":"ErrorResponse","Code":"0A000"}
+{"Type":"ReadyForQuery","TxStatus":"E"}
+
+# Parse should fail in an aborted transaction.
+
+send
+Parse {"Name": "s2", "Query": "SELECT * FROM drop_cols WHERE id = $1"}
+Sync
+----
+
+until
+ErrorResponse
+ReadyForQuery
+----
+{"Type":"ErrorResponse","Code":"25P02"}
+{"Type":"ReadyForQuery","TxStatus":"E"}
+
+# Describe should fail in an aborted transaction.
+
+send
+Describe {"Name": "s1", "ObjectType": "S"}
+Sync
+----
+
+until
+ErrorResponse
+ReadyForQuery
+----
+{"Type":"ErrorResponse","Code":"25P02"}
+{"Type":"ReadyForQuery","TxStatus":"E"}
+
+# Flush should *not* fail in an aborted transaction.
+
+send
+Flush
+Sync
+----
+
+until
+ReadyForQuery
+----
+{"Type":"ReadyForQuery","TxStatus":"E"}
+
+# Bind should fail in an aborted transaction.
+
+send
+Bind {"PreparedStatement": "s1", "Parameters": [{"text": "1"}]}
+Sync
+----
+
+until
+ErrorResponse
+ReadyForQuery
+----
+{"Type":"ErrorResponse","Code":"25P02"}
+{"Type":"ReadyForQuery","TxStatus":"E"}
+
+# Rollback the transaction, and make sure prepared statement works.
+
+send
+Query {"String": "ROLLBACK"}
+Parse {"Name": "s3", "Query": "SELECT * FROM drop_cols WHERE id = $1"}
+Bind {"PreparedStatement": "s3", "Parameters": [{"text": "1"}]}
+Execute
+Sync
+----
+
+until
+ReadyForQuery
+ReadyForQuery
+----
+{"Type":"CommandComplete","CommandTag":"ROLLBACK"}
+{"Type":"ReadyForQuery","TxStatus":"I"}
+{"Type":"ParseComplete"}
+{"Type":"BindComplete"}
+{"Type":"DataRow","Values":[{"text":"1"},{"text":"1"},{"text":"2"}]}
+{"Type":"CommandComplete","CommandTag":"SELECT 1"}
+{"Type":"ReadyForQuery","TxStatus":"I"}
+
+
+# ROLLBACK in a prepared statement is allowed.
+
+send
+Query {"String": "BEGIN"}
+Query {"String": "SELECT 1/0"}
+Parse {"Name": "rollback_stmt", "Query": "ROLLBACK"}
+Bind {"DestinationPortal": "p3", "PreparedStatement": "rollback_stmt"}
+Describe {"Name": "rollback_stmt", "ObjectType": "S"}
+Describe {"Name": "p3", "ObjectType": "P"}
+Execute {"Portal": "p3"}
+Sync
+----
+
+until ignore=RowDescription
+ReadyForQuery
+ErrorResponse
+ReadyForQuery
+----
+{"Type":"CommandComplete","CommandTag":"BEGIN"}
+{"Type":"ReadyForQuery","TxStatus":"T"}
+{"Type":"ErrorResponse","Code":"22012"}
+{"Type":"ReadyForQuery","TxStatus":"E"}
+
+until
+ReadyForQuery
+----
+{"Type":"ParseComplete"}
+{"Type":"BindComplete"}
+{"Type":"ParameterDescription","ParameterOIDs":null}
+{"Type":"NoData"}
+{"Type":"NoData"}
+{"Type":"CommandComplete","CommandTag":"ROLLBACK"}
+{"Type":"ReadyForQuery","TxStatus":"I"}
+
+# ROLLBACK TO SAVEPOINT in a prepared statement is allowed.
+
+send
+Query {"String": "BEGIN"}
+Parse {"Name": "savepoint_stmt", "Query": "SAVEPOINT cockroach_restart"}
+Bind {"DestinationPortal": "p5", "PreparedStatement": "savepoint_stmt"}
+Describe {"Name": "savepoint_stmt", "ObjectType": "S"}
+Describe {"Name": "p5", "ObjectType": "P"}
+Execute {"Portal": "p5"}
+Query {"String": "SELECT 1/0"}
+Parse {"Name": "rollback_savepoint_stmt", "Query": "ROLLBACK TO SAVEPOINT cockroach_restart"}
+Bind {"DestinationPortal": "p6", "PreparedStatement": "rollback_savepoint_stmt"}
+Describe {"Name": "rollback_savepoint_stmt", "ObjectType": "S"}
+Describe {"Name": "p6", "ObjectType": "P"}
+Execute {"Portal": "p6"}
+Sync
+----
+
+until ignore=RowDescription
+ReadyForQuery
+ErrorResponse
+ReadyForQuery
+----
+{"Type":"CommandComplete","CommandTag":"BEGIN"}
+{"Type":"ReadyForQuery","TxStatus":"T"}
+{"Type":"ParseComplete"}
+{"Type":"BindComplete"}
+{"Type":"ParameterDescription","ParameterOIDs":null}
+{"Type":"NoData"}
+{"Type":"NoData"}
+{"Type":"CommandComplete","CommandTag":"SAVEPOINT"}
+{"Type":"ErrorResponse","Code":"22012"}
+{"Type":"ReadyForQuery","TxStatus":"E"}
+
+until
+ReadyForQuery
+----
+{"Type":"ParseComplete"}
+{"Type":"BindComplete"}
+{"Type":"ParameterDescription","ParameterOIDs":null}
+{"Type":"NoData"}
+{"Type":"NoData"}
+{"Type":"CommandComplete","CommandTag":"ROLLBACK"}
+{"Type":"ReadyForQuery","TxStatus":"T"}
+
+# COMMIT in a prepared statement is allowed.
+
+send
+Query {"String": "SELECT 1/0"}
+Parse {"Name": "commit_stmt", "Query": "COMMIT"}
+Bind {"DestinationPortal": "p4", "PreparedStatement": "commit_stmt"}
+Describe {"Name": "commit_stmt", "ObjectType": "S"}
+Describe {"Name": "p4", "ObjectType": "P"}
+Execute {"Portal": "p4"}
+Sync
+----
+
+until ignore=RowDescription
+ErrorResponse
+ReadyForQuery
+----
+{"Type":"ErrorResponse","Code":"22012"}
+{"Type":"ReadyForQuery","TxStatus":"E"}
+
+until
+ReadyForQuery
+----
+{"Type":"ParseComplete"}
+{"Type":"BindComplete"}
+{"Type":"ParameterDescription","ParameterOIDs":null}
+{"Type":"NoData"}
+{"Type":"NoData"}
+{"Type":"CommandComplete","CommandTag":"ROLLBACK"}
+{"Type":"ReadyForQuery","TxStatus":"I"}

